keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,404 @@
+        memory: 256M
+        cpu: "1"
+```
+
+Using the pod-level resource definition enables the `tool1` and `tool2` containers to be constrained by the total limit for the pod. 
+
+Without this feature, the developer would need to decide a-priori how much resources to allocate to the tools - and this is not easy to do for this workload.
+
+### Implementation Details/Notes/Constraints 
+
+The proposal is an **opt-in** feature, and will have no effect on existing deployments. Only deployments that explicitly require this functionality should turn it on by specifying the relevant resource section in the Pod specification.
+
+### Interactions With Other Features
+
+#### Support for cgroup v2
+
+This proposal is compatible with both cgroup v1 and cgroup v2. Both versions of cgroup allow specifying limits on all levels of the cgroup hierarchy.
+
+#### ResourceOverhead
+
+Since the runtime shim is started in the pod-level cgroup, there should be no bad interactions between this proposed feature and the ResourceOverhead feature.
+
+#### Memory-based Emptydir volumes
+
+This proposal doesn't change the current status quo in any way. Memory used for files in a `tmpfs` volume is accounted for as shared-memory. The Linux kernel charges the used memory to cgroup that first touched the shared page for the memory. Moving the charge for shared memory pages between cgroups is not currently supported by the Linux kernel. Defining a memory limit on the pod cgroup level would not change the underlying limitation that memory used by files in the volume are accounted to the cgroup that first touched the memory.
+
+#### HugeTLB cgroup
+
+In progress
+
+### PID, and other cgroups
+
+These are out of scope for this proposal. This proposal doesn't remove the ability to set these limits and requests on the container level, if this was already supported by Kubernetes before.
+
+#### Init Containers
+
+The current implementation (before this proposal) sets up the pod-level cgroup using the maximum between the sum of all the sidecars or the biggest limit for any single Init container. 
+
+There are a few options: 
+1. Limit the pod-level cgroup only after the `InitContainers` have successfully terminated
+1. Require the pod-level cgroup (if it exists) to apply to the `InitContainers` as well.
+1. Allow specifying pod-level restrictions separately for Init containers and sidecars.
+
+Since init containers run sequentially, there is no requirement to share resources with any other container. Therefore the first option is the best - the pod-level cgroup should be restricted only after all of the init containers have finished running.
+
+#### NUMA architectures
+
+There should be no issue caused by limiting a non-leaf cgroup for a pod running in a NUMA environment. The Linux cgroup allows memory from any NUMA node to be used in the same cgroup. See section 5.6 of the Lunux memory cgroup [documentation](https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt).
```

Comment:
Just out of my head, i think there might be some extra complexity required to deal with NUMA, ref. the [topology manager](https://kubernetes.io/docs/tasks/administer-cluster/topology-manager/), with some more context here: https://github.com/kubernetes/enhancements/pull/1752. Maybe some of the people behind that can help with more insight.

Reply:
My (limited) understanding of the NUMA support on Linux is that a set of processes can be restricted via the `cpuset` cgroup to run exclusively on a specific set of CPU nodes and/or memory nodes in the system. Using the `cpuset` cgroup is orthogonal to the `cpu` and `memory` cgroups. If all the containers in a pod fit into a single `cpuset`  (i.e. the topology manager didn't reject the pod), this would mean that the required memory and cpu limit is smaller than the maximum configured amount in the NUMA nodes in that cpuset. Setting the pod-level limits for CPU and memory should therefore be transparent to and non-conflicting with the NUMA configuration.

Could somebody more versed in this topic confirm my understanding?

Reply:
Your understanding is probably correct.

I do however believe the current implementation of the cpu-manager aligns and pins containers separately, while the #1752 add support for doing it on pod-level. If the cpuset is set different on two containers, that might probably make this a bit complecated. So I guess the topology manager has to be aware of this somehow.

An interesting thought is to use this when admitting pods/containers to NUMA-nodes. Eg.: When pod-level resources are set, the topology manager can work on pod level, while it defaults to container otherwise if not set.

For more background info and thoughts:
/cc klueska bg-chun

Reply:
The cpu-manager takes containers in pods that are  `Guaranteed` and pins cpus to them if the limit/requests are integres for the cpu resource. I guess we can just do the same on the pod cgroup for `Guaranteed` pods due to pod limits.